platform/machine/do/cluster.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2017 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package do
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/coreos/pkg/capnslog"
+
+	ctplatform "github.com/coreos/container-linux-config-transpiler/config/platform"
+	"github.com/coreos/mantle/platform"
+	"github.com/coreos/mantle/platform/api/do"
+	"github.com/coreos/mantle/platform/conf"
+)
+
+var (
+	plog = capnslog.NewPackageLogger("github.com/coreos/mantle", "platform/machine/do")
+)
+
+type cluster struct {
+	*platform.BaseCluster
+	api      *do.API
+	sshKeyID int
+}
+
+func NewCluster(opts *do.Options, rconf *platform.RuntimeConfig) (platform.Cluster, error) {
+	api, err := do.New(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	bc, err := platform.NewBaseCluster(opts.BaseName, rconf, ctplatform.DO)
+	if err != nil {
+		return nil, err
+	}
+
+	var keyID int
+	if !rconf.NoSSHKeyInMetadata {
+		keys, err := bc.Keys()
+		if err != nil {
+			return nil, err
+		}
+
+		keyID, err = api.AddKey(context.TODO(), bc.Name(), keys[0].String())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &cluster{
+		BaseCluster: bc,
+		api:         api,
+		sshKeyID:    keyID,
+	}, nil
+}
+
+func (pc *cluster) NewMachine(userdata *conf.UserData) (platform.Machine, error) {
```

Comment:
Would `dc` / `dm` more applicable than `pc` / `pm`?

Reply:
Good catch.  I forgot to update those.